### PR TITLE
feat: allow custom attachments

### PR DESCRIPTION
**Pull request by MartinCupela**
### 🎯 Goal
Allow integrators to add custom attachments to a composed message. This is possible via `upsertAttachment` and `removeAttachment` functions exposed via `MessageInputContextValue`. Both functions expect arrays of values to perform batch updates.

The integrators can customize the look of custom attachments by providing custom components:

1. `UnsupportedAttachment` to `Attachment` component wrapper.
2. `UnknownAttachment` to `AttachmentPreviewList` component wrapper.

`Attachment` and `AttachmentPreviewList` wrappers are then passed to `Channel` as props.

### 🎨 UI Changes

Unknown attachments are by default rendered as follows:

1. Message attachment in a message list:

image

2. Message attachment preview in message composer

image

### CODE REVIEW

File: src/components/MessageInput/hooks/useAttachments.ts [hunk 1]
```diff
@@ -118,39 +151,28 @@ export const useAttachments = <
...
       try {
-        let response: SendFileAPIResponse;
-        if (doFileUploadRequest) {
-          response = await doFileUploadRequest(file, channel);
+        const doUploadRequest = uploadType === 'image' ? doImageUploadRequest : doFileUploadRequest;
+
+        if (doUploadRequest) {
+          response = await doUploadRequest(file, channel);
         } else {
           response = await channel.sendFile(file as File);
```
Comment (MartinCupela) [inline src/components/MessageInput/hooks/useAttachments.ts:174]: call channel.sendImage if uploadType === 'image'

File: src/components/MessageInput/types.ts [hunk 2]
```diff
@@ -91,20 +101,31 @@ export type AudioAttachment<
...
+
 export type AttachmentInternalMetadata = {
   id: string;
   file?: File;
+  previewUri?: string;
   uploadState?: AttachmentLoadingState;
 };
 
-type LocalAttachmentCast<T> = T & { $internal: AttachmentInternalMetadata };
+export type LocalAttachmentCast<T> = T & { $internal: AttachmentInternalMetadata };
```
Comment (MartinCupela) [inline src/components/MessageInput/types.ts:128]: Keep just `LocalAttachment<A,L> => A & {local: LocalMetadata & L }`

```ts
type LocalmageAttachment<
StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = LocalAttachment<ImageAttachment<StreamChatGenerics>, {file: file, previewUri?: string, uploadState?: UploadState}>
```

File: docusaurus/docs/React/guides/message-input/attachment-previews.mdx [hunk 3]
```diff
@@ -0,0 +1,75 @@
+---
+id: attachment_previews
+title: Attachment previews in Message Input
```
Comment (arnautov-anton) [inline docusaurus/docs/React/guides/message-input/attachment-previews.mdx:3]:
```suggestion
title: Attachment Previews in Message Input
```

File: src/components/MessageInput/hooks/useAttachments.ts [hunk 4]
```diff
@@ -89,25 +105,43 @@ export const useAttachments = <
     [maxFilesLeft, noFiles],
   );
 
-  const removeAttachment = (id: string) => {
-    dispatch({ id, type: 'removeAttachment' });
-  };
+  const removeAttachments = useCallback(
+    (ids: string[]) => {
+      if (!(ids && ids.length)) return;
```
Comment (arnautov-anton) [inline src/components/MessageInput/hooks/useAttachments.ts:110]:
```suggestion
      if (!ids.length) return;
```
According to the type annotation `ids` should always be an array so this simple check should suffice.

File: src/components/MessageInput/hooks/useAttachments.ts [hunk 5]
```diff
@@ -89,25 +105,43 @@ export const useAttachments = <
...
+      dispatch({ ids, type: 'removeAttachments' });
+    },
+    [dispatch],
+  );
+
+  const upsertAttachments = useCallback(
+    (attachments: (Attachment<StreamChatGenerics> | LocalAttachment<StreamChatGenerics>)[]) => {
+      if (!(attachments && attachments.length)) return;
```
Comment (arnautov-anton) [inline src/components/MessageInput/hooks/useAttachments.ts:118]:
```suggestion
      if (!attachments.length) return;
```
Same here I suppose. 🤔

File: src/components/MessageInput/hooks/useAttachments.ts [hunk 6]
```diff
@@ -12,15 +12,31 @@ import {
...
 const apiMaxNumberOfFiles = 10;
 
+const ensureIsLocalAttachment = <
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
+>({
+  localMetadata = {},
+  ...attachment
+}:
+  | Attachment<StreamChatGenerics>
+  | LocalAttachment<StreamChatGenerics>): LocalAttachment<StreamChatGenerics> => ({
```
Comment (arnautov-anton) [inline src/components/MessageInput/hooks/useAttachments.ts:32]: Let's check for `localMetadata.id` before re-creating the object - if it already has that established, it should bail.

File: src/components/MessageInput/hooks/useAttachments.ts [hunk 7]
```diff
@@ -164,15 +187,15 @@ export const useAttachments = <
...
           },
-        } as LocalAttachment<StreamChatGenerics>;
+        } as AnyLocalAttachment<StreamChatGenerics>;
 
         dispatch({
-          attachment: failedAttachment,
-          type: 'upsertAttachment',
+          attachments: [failedAttachment],
+          type: 'upsertAttachments',
         });
```
Comment (arnautov-anton) [inline src/components/MessageInput/hooks/useAttachments.ts:199]: Without prior knowledge - could `upsertAttachments` function be used here?

File: src/components/MessageInput/hooks/useMessageInputState.ts [hunk 8]
```diff
@@ -128,16 +130,19 @@ export type MessageInputHookProps<
...
-    attachment: LocalAttachment<StreamChatGenerics>,
-  ) => Promise<LocalAttachment<StreamChatGenerics>>;
+    attachment: AnyLocalAttachment<StreamChatGenerics>,
+  ) => Promise<AnyLocalAttachment<StreamChatGenerics> | undefined>;
   uploadFile: (id: string) => void;
   uploadImage: (id: string) => void;
   uploadNewFiles: (files: FileList | File[]) => void;
+  upsertAttachments: (
+    attachments: (Attachment<StreamChatGenerics> | LocalAttachment<StreamChatGenerics>)[],
+  ) => void;
```
Comment (arnautov-anton) [inline src/components/MessageInput/hooks/useMessageInputState.ts:145]:
```suggestion
  upsertAttachments: (
    attachments: (Attachment<StreamChatGenerics> | LocalAttachment<StreamChatGenerics>)[],
  ) => void;
```
nit/feat: Could integrators benefit from patch function?

```ts
upsertAttachments((currentAttachments) => {
  return currentAttachments.map((attachment) => {
    if (attachment.id !== arbitraryId) return attachment;
    return { ...attachment, localMetadata: { ...attachment.localMetadata, customThing: 123 } };
  });
});
```

File: src/components/MessageInput/hooks/useAttachments.ts (hunk 4 above)
Comment (MartinCupela) [reply src/components/MessageInput/hooks/useAttachments.ts:110]: I was too paranoid :)

File: src/components/MessageInput/hooks/useAttachments.ts (hunk 6 above)
Comment (MartinCupela) [reply src/components/MessageInput/hooks/useAttachments.ts:32]: adjusted

File: src/components/MessageInput/hooks/useAttachments.ts (hunk 5 above)
Comment (MartinCupela) [reply src/components/MessageInput/hooks/useAttachments.ts:118]: adjusted

File: src/components/MessageInput/hooks/useAttachments.ts (hunk 4 above)
Comment (MartinCupela) [reply src/components/MessageInput/hooks/useAttachments.ts]: adjusted

File: src/components/MessageInput/hooks/useAttachments.ts (hunk 7 above)
Comment (MartinCupela) [reply src/components/MessageInput/hooks/useAttachments.ts]: yes, adjusted

File: src/components/MessageInput/hooks/useMessageInputState.ts [hunk 9]
```diff
@@ -271,31 +282,35 @@ const messageInputReducer = <
...
+      action.attachments.forEach((actionAttachment) => {
+        const attachmentIndex = state.attachments.findIndex(
+          (att) =>
+            att.localMetadata?.id && att.localMetadata?.id === actionAttachment.localMetadata?.id,
+        );
+
+        if (attachmentIndex === -1) {
+          attachments.push(actionAttachment);
+        } else {
+          const upsertedAttachment = mergeDeep(
```
Comment (MartinCupela) [inline src/components/MessageInput/hooks/useMessageInputState.ts:296]: @arnautov-anton regarding your question about patching the attachments. It happens here.

**Review by arnautov-anton — APPROVED** [review]
🌮

**Pull request merged**